File: modules/core/src/encoder/segment/PublisherTCEncoder.ts
import {

  SegmentType,

} from '.';

import {

  Base64Url,
  BitLength,

} from '../';

import {

  FieldEncoderMap,
  IntEncoder,

} from '../field';

import {

  PublisherFieldSequence,

} from '../sequence';

import {

  TCModel,
  TCModelPropType,

} from '../../';

export class PublisherTCEncoder {

  public static encode(tcModel: TCModel): string {

    const encMap: FieldEncoderMap = new FieldEncoderMap();
    const pubFieldSequence: PublisherFieldSequence = new PublisherFieldSequence();
    const encodeSequence: string[] = pubFieldSequence[tcModel.version.toString()];

    // first encode the segment type
    let bitField: string = IntEncoder.encode(SegmentType.publisherTC, BitLength.segmentType);

    encodeSequence.forEach((key: string): void => {

      const value: TCModelPropType = tcModel[key];
      const encoder = encMap[key];
      const numBits: number = (BitLength[key]) ? BitLength[key] : tcModel.numCustomPurposes;

      bitField += encoder.encode(value, numBits);

    });

    // base64url encode the string and return
    return Base64Url.encode(bitField);

  }

  public static decode(encodedString: string, tcModel: TCModel): TCModel {

    const encMap: FieldEncoderMap = new FieldEncoderMap();
    const pubFieldSequence: PublisherFieldSequence = new PublisherFieldSequence();
    const encodeSequence: string[] = pubFieldSequence[tcModel.version.toString()];
    let bStringIdx = 0;
    let bitField = Base64Url.decode(encodedString);

    /**
     * the first n bits are the type we don't actually care about them because
     * it's handled at a different level
     */
    bitField = bitField.substr(BitLength.segmentType);

    encodeSequence.forEach((key: string): void => {

      const encoder = encMap[key];
      const numBits: number = (BitLength[key]) ? BitLength[key] : tcModel.numCustomPurposes;

      tcModel[key] = encoder.decode(bitField.substr(bStringIdx, numBits));
      bStringIdx += numBits;

    });

    return tcModel;

  }

}

File: GDPR-Phase2-Java/src/main/java/com/gdpr/phase2/encoder/segment/PublisherTCEncoder.java
package com.gdpr.phase2.encoder.segment;

import com.gdpr.phase2.encoder.Base64Url;
import com.gdpr.phase2.encoder.sequence.PublisherFieldSequence;
import com.gdpr.phase2.Utils.TCModel;
import com.gdpr.phase2.Utils.TCModelEnum;
import com.gdpr.phase2.encoder.BaseEncoder;
import com.gdpr.phase2.encoder.BitLength;
import com.gdpr.phase2.encoder.field.FieldEncoderMap;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

public class PublisherTCEncoder implements BaseSegmentEncoder {
    private PublisherTCEncoder() {
    }
    private static final PublisherTCEncoder instance = new PublisherTCEncoder();
    public static PublisherTCEncoder getInstance() {
        return instance;
    }

    public TCModel decode(String encodedString, TCModel tcModel) {
        Map<String, BaseEncoder> encMap = FieldEncoderMap.getInstance().fieldMap;
        PublisherFieldSequence pubFieldSequence = PublisherFieldSequence.getInstance();
        List<String> encodedSequence = null;
        if(tcModel.getVersion() == 1) {
            encodedSequence = pubFieldSequence.one;
        } else {
            encodedSequence = pubFieldSequence.two;
        }
        AtomicInteger bStringIdx = new AtomicInteger();
        String bitField = Base64Url.decode(encodedString);
        encodedSequence.forEach(key -> {
            BaseEncoder encoder = encMap.get(key);
            Object keyLength = null;
            keyLength = Optional.ofNullable(BitLength.fieldLengths.get(key)).orElse(0);
            int numBits = (int) keyLength > 0 ? (int)keyLength : tcModel.getNumCustomPurposes();
            TCModelEnum.valueOf(key).setValue(tcModel,encoder.decode(bitField.substring(bStringIdx.get(), bStringIdx.get() + numBits)));
            bStringIdx.addAndGet(numBits);
        });
        return tcModel;
    }
}
